admin/aws.py

```diff
@@ -380,11 +425,31 @@ def __new__(cls, value, content_type):
         return self
 
 
-@attributes([
-    Attribute('routing_rules'),
-    Attribute('s3_buckets'),
-    Attribute('error_key', default_factory=dict)
-])
+class FakeAWSState(PClass):
+    """
+    The immutable state of ``FakeAWS``
+
+    :ivar routing_rules: Dictionary of routing rules for S3 buckets. They are
+        represented as dictonaries mapping key prefixes to replacements. Other
+        types of rules and attributes are supported or represented.
+    :ivar s3_buckets: Dictionary of fake S3 buckets. Each bucket is represented
+        as a dictonary mapping keys to contents. Other attributes are ignored.
+    :ivar cloudfront_invalidations: List of
+        :class:`CreateCloudFrontInvalidation` that have been requested.
+    """
+    routing_rules = pmap_field(
+        key_type=unicode,
+        value_type=RoutingRules
+    )
+    s3_buckets = pmap_field(
+        key_type=unicode,
+        value_type=PMap
+    )
+    error_key = pmap_field(key_type=unicode, value_type=unicode)
+    cloudfront_invalidations = field(initial=freeze([]))
```

Comment:
:exclamation: If you used more specific types for all these, you woudn't need to litter the test code with tons of `freeze` calls, which are just noise.